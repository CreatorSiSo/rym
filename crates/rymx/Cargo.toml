[package]
name = "rymx"
version = "0.0.3"
edition = "2021"
default-run = "rymx"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
anyhow = "1.0"
clap = { version = "4.4", features = ["cargo"] }
insta = { version = "1.31.0", features = ["glob"] }
logos = "0.13.0"
owo-colors = "3.5"
rustyline = "12"

[features]
expand = []

[profile.dev.package.insta]
opt-level = 3

[profile.dev.package.similar]
opt-level = 3
