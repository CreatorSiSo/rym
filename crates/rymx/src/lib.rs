use std::fmt::Write;

mod ast;
mod compile;
mod diagnostics;
mod interpret;
mod parse;
mod span;
pub mod std_lib;
mod tokenize;

pub use diagnostics::{Diagnostics, SourceId};
pub use interpret::Env;
pub use span::Span;
pub use tokenize::tokenizer;

use ast::{Expr, Module};
use interpret::{Interpret, Value};
use span::SourceSpan;
use tokenize::Token;

pub fn interpret(diag: &mut Diagnostics, env: &mut Env, ast: impl Interpret) -> Value {
	diag.start_stage("interpret");
	// TODO does this make sense
	// Ignoring control flow
	let result = ast.eval(env).inner();

	let env_state: String = env
		.variables()
		.into_iter()
		.fold((0, String::new()), |(indent, mut accum), scope| {
			for (name, (kind, value)) in scope {
				writeln!(accum, "{}{kind} {name} = {}", "  ".repeat(indent), value)
					.expect("Internal Error: Unable to write into String");
			}
			(indent + 1, accum)
		})
		.1;
	diag.push_result(&env_state);

	result
}

pub fn compile_module(diag: &mut Diagnostics, src: &str, src_id: SourceId) -> Option<Module> {
	diag.start_stage("tokenize");
	let tokens: Vec<(Token, Span)> = tokenize(diag, src, src_id.clone())?;

	diag.start_stage("parse");
	let module = match parse::parse_file(&tokens, src, src_id) {
		Ok(module) => module,
		Err(reports) => {
			for report in reports {
				diag.push_report(report);
			}
			return None;
		}
	};
	diag.push_result(&format!("{module:#?}\n"));

	// TODO Name resolution
	// TODO Typechecking
	// TODO Const evaluation
	// TODO Generate intermediate representation

	// i / / / ;

	Some(module)
}

// TODO take a module (for name lookup and so on) as input
pub fn compile_expr(diag: &mut Diagnostics, src: &str, src_id: SourceId) -> Option<Expr> {
	diag.start_stage("tokenize");
	let tokens: Vec<(Token, Span)> = tokenize(diag, src, src_id.clone())?;

	diag.start_stage("parse");
	let expr = match parse::parse_expr(&tokens, src, src_id) {
		Ok(expr) => expr,
		Err(reports) => {
			for report in reports {
				diag.push_report(report);
			}
			return None;
		}
	};
	diag.push_result(&format!("{expr:#?}\n"));

	Some(expr)
}

fn tokenize(diag: &mut Diagnostics, src: &str, src_id: SourceId) -> Option<Vec<(Token, Span)>> {
	let results: Vec<_> = tokenizer(src).collect();
	diag.push_result(
		&results.iter().fold(String::new(), |accum, (result, span)| {
			let (token, span) = match result {
				Ok(token) => (format!("{token:?}"), span),
				Err(_) => ("Error".into(), span),
			};
			accum + &format!("{token} [{}]\n", span.src(src).escape_debug())
		}),
	);

	let mut tokens = vec![];
	for (result, span) in results {
		let Ok(token) = result else {
			diag.error(
				format!("Invalid character [{}]", span.src(src)),
				SourceSpan(src_id, span),
			);
			return None;
		};
		match token {
			Token::DocComment | Token::Comment | Token::VSpace | Token::HSpace => continue,
			_ => tokens.push((token, span)),
		}
	}
	Some(tokens)
}
