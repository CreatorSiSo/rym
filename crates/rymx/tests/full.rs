use insta::assert_snapshot;
use rymx::Diagnostics;

#[test]
fn main() {
	insta::glob!("**/*.rym", |path| {
		let src = std::fs::read_to_string(path).unwrap();
		let mut diag = Diagnostics::new(Some(path.into()), Box::new(std::io::stderr()));
		let _ = rymx::compile_module(&mut diag, &src);
		assert_snapshot!(diag.dump_stages());
	})
}
