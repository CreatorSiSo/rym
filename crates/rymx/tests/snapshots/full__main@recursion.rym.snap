---
source: crates/rymx/tests/full.rs
expression: diag.outputs_dump() + &diag.reports_dump()
input_file: crates/rymx/tests/recursion.rym
---
--- tokenize ---
Fn [fn]
HSpace [ ]
Ident [main]
ParenOpen [(]
ParenClose [)]
HSpace [ ]
Ident [void]
HSpace [ ]
BraceOpen [{]
VSpace [\n]
HSpace [    ]
Ident [test_fib]
ParenOpen [(]
String [\"fib_if\"]
Comma [,]
HSpace [ ]
Ident [fib_if]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Comment [// test_fib(\"fib_match\", fib_match);]
VSpace [\n]
HSpace [    ]
Comment [// test_fib(\"fib_oneliner\", fib_oneliner);]
VSpace [\n\n]
HSpace [    ]
Ident [test_countdown]
ParenOpen [(]
ParenClose [)]
Semi [;]
VSpace [\n]
BraceClose [}]
VSpace [\n\n]
Fn [fn]
HSpace [ ]
Ident [fib_if]
ParenOpen [(]
Ident [n]
ParenClose [)]
HSpace [ ]
Ident [Uint]
VSpace [\n]
HSpace [    ]
If [if]
HSpace [ ]
Ident [n]
HSpace [ ]
LessThanEq [<=]
HSpace [ ]
Int [0]
HSpace [ ]
Then [then]
HSpace [ ]
Int [0]
VSpace [\n]
HSpace [    ]
Else [else]
HSpace [ ]
If [if]
HSpace [ ]
Ident [n]
HSpace [ ]
Eq [==]
HSpace [ ]
Int [1]
HSpace [ ]
Then [then]
HSpace [ ]
Int [1]
VSpace [\n]
HSpace [    ]
Else [else]
HSpace [ ]
Ident [fib_if]
ParenOpen [(]
Ident [n]
HSpace [ ]
Minus [-]
HSpace [ ]
Int [1]
ParenClose [)]
HSpace [ ]
Plus [+]
HSpace [ ]
Ident [fib_if]
ParenOpen [(]
Ident [n]
HSpace [ ]
Minus [-]
HSpace [ ]
Int [2]
ParenClose [)]
VSpace [\n\n]
Comment [// const fib_match = fn(n) => match n with]
VSpace [\n]
Comment [//     | 0 => 0]
VSpace [\n]
Comment [//     | 1 => 1]
VSpace [\n]
Comment [//     | _ => fib_match(n - 1) + fib_match(n - 2);]
VSpace [\n\n]
Comment [// const fib_oneliner = fn(n) => match n with 0 => 0 | 1 => 1 | _ => fib_oneliner(n - 1) + fib_oneliner(n - 2);]
VSpace [\n\n]
Fn [fn]
HSpace [ ]
Ident [test_fib]
ParenOpen [(]
Ident [name]
Comma [,]
HSpace [ ]
Ident [fib]
ParenClose [)]
HSpace [ ]
Ident [void]
HSpace [ ]
BraceOpen [{]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [0]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [0]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [1]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [1]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [2]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [1]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [3]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [2]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [4]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [3]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [5]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [5]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [6]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [8]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [7]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [13]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [8]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [21]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [9]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [34]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [10]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [55]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [assert_eq]
ParenOpen [(]
Ident [fib]
ParenOpen [(]
Int [20]
ParenClose [)]
Comma [,]
HSpace [ ]
Int [6765]
ParenClose [)]
Semi [;]
VSpace [\n\n]
HSpace [    ]
Ident [println]
ParenOpen [(]
String [\"Successfully tested \"]
Comma [,]
HSpace [ ]
Ident [name]
Comma [,]
HSpace [ ]
String [\"!\"]
ParenClose [)]
Semi [;]
VSpace [\n]
BraceClose [}]
VSpace [\n\n]
Fn [fn]
HSpace [ ]
Ident [countdown]
ParenOpen [(]
Ident [n]
Comma [,]
HSpace [ ]
Ident [step]
ParenClose [)]
HSpace [ ]
Ident [void]
HSpace [ ]
BraceOpen [{]
VSpace [\n]
HSpace [    ]
Ident [print]
ParenOpen [(]
Ident [n]
Comma [,]
HSpace [ ]
String [\" \"]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
If [if]
HSpace [ ]
Ident [n]
HSpace [ ]
GreaterThan [>]
HSpace [ ]
Int [0]
HSpace [ ]
Then [then]
HSpace [ ]
Ident [countdown]
ParenOpen [(]
Ident [min]
ParenOpen [(]
Ident [n]
HSpace [ ]
Minus [-]
HSpace [ ]
Ident [step]
Comma [,]
HSpace [ ]
Int [0]
ParenClose [)]
Comma [,]
HSpace [ ]
Ident [step]
ParenClose [)]
Semi [;]
VSpace [\n]
BraceClose [}]
VSpace [\n\n]
Fn [fn]
HSpace [ ]
Ident [min]
ParenOpen [(]
Ident [value]
Comma [,]
HSpace [ ]
Ident [min]
ParenClose [)]
HSpace [ ]
Ident [any]
VSpace [\n]
HSpace [    ]
If [if]
HSpace [ ]
Ident [value]
HSpace [ ]
LessThan [<]
HSpace [ ]
Ident [min]
HSpace [ ]
Then [then]
HSpace [ ]
Ident [min]
HSpace [ ]
Else [else]
HSpace [ ]
Ident [value]
VSpace [\n\n]
Fn [fn]
HSpace [ ]
Ident [max]
ParenOpen [(]
Ident [value]
Comma [,]
HSpace [ ]
Ident [max]
ParenClose [)]
HSpace [ ]
Ident [any]
VSpace [\n]
HSpace [    ]
If [if]
HSpace [ ]
Ident [value]
HSpace [ ]
GreaterThan [>]
HSpace [ ]
Ident [max]
HSpace [ ]
Then [then]
HSpace [ ]
Ident [max]
HSpace [ ]
Else [else]
HSpace [ ]
Ident [value]
VSpace [\n\n]
Fn [fn]
HSpace [ ]
Ident [clamp]
ParenOpen [(]
Ident [value]
Comma [,]
HSpace [ ]
Ident [min]
Comma [,]
HSpace [ ]
Ident [max]
ParenClose [)]
HSpace [ ]
Ident [any]
VSpace [\n]
HSpace [    ]
If [if]
HSpace [ ]
Ident [value]
HSpace [ ]
LessThan [<]
HSpace [ ]
Ident [min]
HSpace [ ]
Then [then]
VSpace [\n]
HSpace [        ]
Ident [min]
VSpace [\n]
HSpace [    ]
Else [else]
HSpace [ ]
If [if]
HSpace [ ]
Ident [value]
HSpace [ ]
GreaterThan [>]
HSpace [ ]
Ident [max]
HSpace [ ]
Then [then]
VSpace [\n]
HSpace [        ]
Ident [max]
VSpace [\n]
HSpace [    ]
Else [else]
VSpace [\n]
HSpace [        ]
Ident [value]
VSpace [\n\n]
Fn [fn]
HSpace [ ]
Ident [test_countdown]
ParenOpen [(]
ParenClose [)]
HSpace [ ]
Ident [void]
HSpace [ ]
BraceOpen [{]
VSpace [\n]
HSpace [    ]
Ident [countdown]
ParenOpen [(]
Int [0]
Comma [,]
HSpace [ ]
Int [5]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [println]
ParenOpen [(]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [countdown]
ParenOpen [(]
Int [1]
Comma [,]
HSpace [ ]
Int [1]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [println]
ParenOpen [(]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [countdown]
ParenOpen [(]
Int [10]
Comma [,]
HSpace [ ]
Int [1]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [println]
ParenOpen [(]
ParenClose [)]
Semi [;]
VSpace [\n]
HSpace [    ]
Ident [countdown]
ParenOpen [(]
Int [100]
Comma [,]
HSpace [ ]
Int [8]
ParenClose [)]
Semi [;]
VSpace [\n]
BraceClose [}]
VSpace [\n]

--- parse ---
Module {
    name: "",
    constants: [
        (
            "main",
            Function {
                name: Some(
                    "main",
                ),
                params: [],
                return_type: Some(
                    "void",
                ),
                body: Block(
                    [
                        Call(
                            Ident("test_fib"),
                            [
                                Literal(String fib_if),
                                Ident("fib_if"),
                            ],
                        ),
                        Call(
                            Ident("test_countdown"),
                            [],
                        ),
                    ],
                ),
            },
        ),
        (
            "fib_if",
            Function {
                name: Some(
                    "fib_if",
                ),
                params: [
                    (
                        "n",
                        (),
                    ),
                ],
                return_type: Some(
                    "Uint",
                ),
                body: IfElse(
                    LessThanEq(
                        Ident("n"),
                        Literal(Int 0),
                    ),
                    Literal(Int 0),
                    IfElse(
                        Eq(
                            Ident("n"),
                            Literal(Int 1),
                        ),
                        Literal(Int 1),
                        Add(
                            Call(
                                Ident("fib_if"),
                                [
                                    Sub(
                                        Ident("n"),
                                        Literal(Int 1),
                                    ),
                                ],
                            ),
                            Call(
                                Ident("fib_if"),
                                [
                                    Sub(
                                        Ident("n"),
                                        Literal(Int 2),
                                    ),
                                ],
                            ),
                        ),
                    ),
                ),
            },
        ),
        (
            "test_fib",
            Function {
                name: Some(
                    "test_fib",
                ),
                params: [
                    (
                        "name",
                        (),
                    ),
                    (
                        "fib",
                        (),
                    ),
                ],
                return_type: Some(
                    "void",
                ),
                body: Block(
                    [
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 0),
                                    ],
                                ),
                                Literal(Int 0),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 1),
                                    ],
                                ),
                                Literal(Int 1),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 2),
                                    ],
                                ),
                                Literal(Int 1),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 3),
                                    ],
                                ),
                                Literal(Int 2),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 4),
                                    ],
                                ),
                                Literal(Int 3),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 5),
                                    ],
                                ),
                                Literal(Int 5),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 6),
                                    ],
                                ),
                                Literal(Int 8),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 7),
                                    ],
                                ),
                                Literal(Int 13),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 8),
                                    ],
                                ),
                                Literal(Int 21),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 9),
                                    ],
                                ),
                                Literal(Int 34),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 10),
                                    ],
                                ),
                                Literal(Int 55),
                            ],
                        ),
                        Call(
                            Ident("assert_eq"),
                            [
                                Call(
                                    Ident("fib"),
                                    [
                                        Literal(Int 20),
                                    ],
                                ),
                                Literal(Int 6765),
                            ],
                        ),
                        Call(
                            Ident("println"),
                            [
                                Literal(String Successfully tested ),
                                Ident("name"),
                                Literal(String !),
                            ],
                        ),
                    ],
                ),
            },
        ),
        (
            "countdown",
            Function {
                name: Some(
                    "countdown",
                ),
                params: [
                    (
                        "n",
                        (),
                    ),
                    (
                        "step",
                        (),
                    ),
                ],
                return_type: Some(
                    "void",
                ),
                body: Block(
                    [
                        Call(
                            Ident("print"),
                            [
                                Ident("n"),
                                Literal(String  ),
                            ],
                        ),
                        IfElse(
                            GreaterThan(
                                Ident("n"),
                                Literal(Int 0),
                            ),
                            Call(
                                Ident("countdown"),
                                [
                                    Call(
                                        Ident("min"),
                                        [
                                            Sub(
                                                Ident("n"),
                                                Ident("step"),
                                            ),
                                            Literal(Int 0),
                                        ],
                                    ),
                                    Ident("step"),
                                ],
                            ),
                            Unit,
                        ),
                    ],
                ),
            },
        ),
        (
            "min",
            Function {
                name: Some(
                    "min",
                ),
                params: [
                    (
                        "value",
                        (),
                    ),
                    (
                        "min",
                        (),
                    ),
                ],
                return_type: Some(
                    "any",
                ),
                body: IfElse(
                    LessThan(
                        Ident("value"),
                        Ident("min"),
                    ),
                    Ident("min"),
                    Ident("value"),
                ),
            },
        ),
        (
            "max",
            Function {
                name: Some(
                    "max",
                ),
                params: [
                    (
                        "value",
                        (),
                    ),
                    (
                        "max",
                        (),
                    ),
                ],
                return_type: Some(
                    "any",
                ),
                body: IfElse(
                    GreaterThan(
                        Ident("value"),
                        Ident("max"),
                    ),
                    Ident("max"),
                    Ident("value"),
                ),
            },
        ),
        (
            "clamp",
            Function {
                name: Some(
                    "clamp",
                ),
                params: [
                    (
                        "value",
                        (),
                    ),
                    (
                        "min",
                        (),
                    ),
                    (
                        "max",
                        (),
                    ),
                ],
                return_type: Some(
                    "any",
                ),
                body: IfElse(
                    LessThan(
                        Ident("value"),
                        Ident("min"),
                    ),
                    Ident("min"),
                    IfElse(
                        GreaterThan(
                            Ident("value"),
                            Ident("max"),
                        ),
                        Ident("max"),
                        Ident("value"),
                    ),
                ),
            },
        ),
        (
            "test_countdown",
            Function {
                name: Some(
                    "test_countdown",
                ),
                params: [],
                return_type: Some(
                    "void",
                ),
                body: Block(
                    [
                        Call(
                            Ident("countdown"),
                            [
                                Literal(Int 0),
                                Literal(Int 5),
                            ],
                        ),
                        Call(
                            Ident("println"),
                            [],
                        ),
                        Call(
                            Ident("countdown"),
                            [
                                Literal(Int 1),
                                Literal(Int 1),
                            ],
                        ),
                        Call(
                            Ident("println"),
                            [],
                        ),
                        Call(
                            Ident("countdown"),
                            [
                                Literal(Int 10),
                                Literal(Int 1),
                            ],
                        ),
                        Call(
                            Ident("println"),
                            [],
                        ),
                        Call(
                            Ident("countdown"),
                            [
                                Literal(Int 100),
                                Literal(Int 8),
                            ],
                        ),
                    ],
                ),
            },
        ),
    ],
    children: [],
}

--- interpret ---
const assert = fn<native> (1)
const assert_eq = fn<native> (2)
const clamp = fn (3)
const countdown = fn (2)
const fib_if = fn (1)
const fib_native = fn<native> (1)
const main = fn (0)
const max = fn (2)
const min = fn (2)
const print = fn<native> (variadic)
const println = fn<native> (variadic)
const read_to_string = fn<native> (1)
const test_countdown = fn (0)
const test_fib = fn (2)

--- reports ---

